### ISSUE TITLE
Discussion of CPO push mechanism

### ISSUE THREAD

**Issue by julius-yang** [ocpi/ocpi#460]
The 2.x.x spec allows for CPOs to push data to eMSPs, or for eMSPs to pull from CPOs.  I'd like to know from other eMSPs what their experience with push/pull is.  Here's our experience with 2.x.x:

We are an eMSP.  Some CPOs push data to us, and some ask us to pull.  For the ones that push, we set up a "baseline sync" which is a periodic pull to ensure our data is up to date.  The baseline sync is a pull of the entire CPO data set spread over the course of about a day.

There are drawbacks to both approaches.

**eMSP pulls from CPO**
The quantity of data is so vast that the CPO cannot serve all the data rapidly enough (e.g. for EVSE status)
The amount of data to process is so vast that the eMSP service is overwhelmed

**CPO pushes to eMSP**
OCPP/hardware/firmware bugs transmit nonstop status changes to OCPI, which pushes to the eMSP
eMSP has no control over quantity of data received -- it has to severely overprovision services or otherwise rate-limit to protect against flooding
CPO has to push to N different eMSPs

I feel that the advantage of an eMSP pulling from a CPO is large:

1. CPO can cache data for eMSPs, increasing throughput drastically for the eMSP
2. eMSP can control rate of ingestion, protecting themselves
3. If there is a service outage on either side, the eMSP can set time-sensitive values such as status to an unknown state; the next pull will true up
4. CPO no longer has to trigger a push to every eMSP; just write to the cache and let the eMSP get it

But as I have not worked on the CPO side I'm interested in hearing what other people think.  My gut feeling is that CPO push to eMSP is better for small CPOs, but at scale eMSP pulls from the CPO is far better.

**Comment by ahzf** [comment]
Yes..  for many EMSPs and CPOs Hubject and OICP is the far better option ;) With OICP you could still do some messages p2p, even when Hubject does not describe this operation mode of OICP themselves... but technical this is possible.

For all others you have to deal with the problem, that within a p2p architecture all participants have to survive the load coming from the biggest misbehaving partner system. That is the dark side of p2p ;)

A pulling-EMSP is in my eyes a "broken" EMSP, as e.g. status updates have to be near real time... also CPOs should aggregate e.g. all status changes for let's say 5 seconds an push a list of status updates afterwards. The current way OCPI handles status updates is really bad for everyone.

**Comment by ev-adias** [comment]
We do both CPO and EMSP so we don't have a problem with badly behaving actors, but a common way to deal with big amounts of "push messages" is to put a queue.  Just accept everything put it in a queue and process it at the rhythm that you can handle.  It introduces some delay, but it's better than breaking under load.
Honestly, although HTTP is almost always the best option, a protocol more suited to events/messages would help a lot. A publish-subscribe protocol could even remove the need for intermediaries.
